File: src/Enterprise/VisualBasic/Example1.vb
Imports System

' Abstract Product
Public Interface Button
    Sub Render()
End Interface

Public Interface Checkbox
    Sub Render()
End Interface

' Concrete Product
Public Class DarkButton
    Implements Button
    Public Sub Render() Implements Button.Render
        Console.WriteLine("Dark Button")
    End Sub
End Class

Public Class LightButton
    Implements Button
    Public Sub Render() Implements Button.Render
        Console.WriteLine("Light Button")
    End Sub
End Class

Public Class DarkCheckbox
    Implements Checkbox
    Public Sub Render() Implements Checkbox.Render
        Console.WriteLine("Dark Checkbox")
    End Sub
End Class

Public Class LightCheckbox
    Implements Checkbox
    Public Sub Render() Implements Checkbox.Render
        Console.WriteLine("Light Checkbox")
    End Sub
End Class

' Abstract Factory
Public Interface UIFactory
    Function CreateButton() As Button
    Function CreateCheckbox() As Checkbox
End Interface

' Concrete Factory
Public Class DarkFactory
    Implements UIFactory
    Public Function CreateButton() As Button Implements UIFactory.CreateButton
        Return New DarkButton()
    End Function
    Public Function CreateCheckbox() As Checkbox Implements UIFactory.CreateCheckbox
        Return New DarkCheckbox()
    End Function
End Class

Public Class LightFactory
    Implements UIFactory
    Public Function CreateButton() As Button Implements UIFactory.CreateButton
        Return New LightButton()
    End Function
    Public Function CreateCheckbox() As Checkbox Implements UIFactory.CreateCheckbox
        Return New LightCheckbox()
    End Function
End Class

Module Example1
    Sub CreateUIComponents(factory As UIFactory)
        Dim button As Button = factory.CreateButton()
        Dim checkbox As Checkbox = factory.CreateCheckbox()
        button.Render()
        checkbox.Render()
    End Sub

    Sub Main()
        CreateUIComponents(New DarkFactory())
        CreateUIComponents(New LightFactory())
    End Sub
End Module

File: src/Niche/Prolog/example1.pl
% Abstract Factory
%
% A factory term represents one coherent product family. The consumer selects
% the family once and receives both products from that same factory.

ui_factory(dark, factory(dark_button, dark_checkbox)).
ui_factory(light, factory(light_button, light_checkbox)).

create_ui_components(factory(Button, Checkbox)) :-
    call(Button),
    call(Checkbox).

% Concrete Products
dark_button :- write('Dark Button'), nl.
light_button :- write('Light Button'), nl.

dark_checkbox :- write('Dark Checkbox'), nl.
light_checkbox :- write('Light Checkbox'), nl.

% Usage
run :-
    ui_factory(dark, Factory),
    create_ui_components(Factory).
